### DISCUSSION TITLE
Node-level regression problem: take inputs of a subset of nodes to predict the behavior of all nodes

### DISCUSSION THREAD

**Discussion by DrKleiber** [pyg-team/pytorch_geometric#4493] [Q&A] [answered]
I'm new to graph models and am working on a problem for energy system prediction. I model the system as a graph with each node representing a certain component. I can place sensors in certain nodes but not all. So in real life, I can only know the behavior of a subset of nodes of the graph through sensor measurements. 

My goal is to train a (static) graph model that reads in the sensor measurements of a subset of nodes and predicts the behavior of all nodes. To do that, I ran a few hundred simulations and get the behavior of all nodes under certain conditions.  I want to perform the training based on these results, so the model can be used for unseen conditions. But I didn't find a similar example to my problem.

I read the Colab tutorial for node classification, where we define train_mask so that model reads in the features of all the nodes and calculates the error function based on the labels of a subset of nodes. Looks like contrary to my case, but when I tried to apply the train_mask to input( data.x[data.train_mask]), error showed up, as looks like x[data.train_mask] is inconsistent with the given graph. I wonder is there any possible solution to my problem? Would padding the unknown node with zero in the input be a viable option?

Any suggestions/examples would be highly appreciated! Thank you in advance!

**Comment by rusty1s** [comment, marked as answer]
Shapes of `x` and `train_mask` need to match in the first dimension, that is, the both should hold values for every node in the graph. If you only want to learn on a subgraph but apply the model on the full graph later on, how about you generate your sub-data via
```python
train_data = data.subgraph(train_mask_or_indices)
```
and use it to train your model in a fully inductive mode. WDYT?

**Reply by DrKleiber** [reply]
Thank you @rusty1s for your prompt response. One thing I want to clarify is I want my input to be a subgraph, but the output is the full graph. Is the method you mentioned still workable? Anyway I'll start trying this subgraph approach. Thank you!

**Reply by rusty1s** [reply]
Mh, that is a completely different problem, and would correspond to a graph generation problem. Is that really what you are trying to achieve here? I thought you want to train your model on a subset of the data, and then apply it on unseen graphs.

**Reply by DrKleiber** [reply]
Hi @rusty1s, sorry I didn't express my question clearly and caused confusion. To put it simple, I have a known static graph with four nodes A,B, C, D, their connections are fixed. I always know the attributes of Node A. I know the attributes of other nodes for some cases, but not for unseen cases. I want to train the graph neural network so that I input whatever I observed in A, the model can predict the attributes in B, C, and D.  
My understanding is from the graph neural network perspective, my problem is not that complicated (I definitely do not need to generate graph, as the graph structure in my problem is known and fixed, I just don't know the attributes of some of the nodes). Certainly, the real problem is more complicated than the toy example I mentioned and I need to involve heterogeneous graph, and for temporal prediction.

**Reply by rusty1s** [reply]
This looks like a classical masked autoencoder approach. That is, you input your graph and features into the GNN, where the features for unknown nodes are set to zero. After message passing completes, you try to reconstruct the features of the unknown nodes (using MSE loss):
```python
train_mask = ... # zero for nodes we know the features of.

x = data.x.clone()
x[~train_mask] = 0.
out = model(x, data.edge_index)  # Maps to the same shape as `data.x`
loss = F.mse_loss(out[~train_mask], data.x[~train_mask])
loss.backward()
```

**Reply by DrKleiber** [reply]
Thank you @rusty1s ! I'll try it.

**Reply by atello** [reply]
Hi @rusty1s, 

First of all,  thanks for helping us, who are starting with GNNs, with your guidance. 
I am working on a similar setting than the one that is described here. Just to be sure, in the line where you calculate the **mse_loss** shouldn't it be **out**[~train_mask] instead of **x**[~train_mask]

`loss = F.mse_loss(out[~train_mask], data.x[~train_mask])`

**Reply by rusty1s** [reply]
Yes, absolutely. Thanks for pointing that out. Edited my code snippet.

**Reply by atello** [reply]
Thanks for your reply @rusty1s!
